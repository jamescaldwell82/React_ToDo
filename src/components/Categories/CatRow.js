import React, { useState } from 'react'
import axios from 'axios'

import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { library } from '@fortawesome/fontawesome-svg-core';
import { fas } from '@fortawesome/free-solid-svg-icons';
library.add(fas);

export default function CatRow(props) {
    const [editMode, setEditMode] = useState(false);
    const [name, setName] = useState(props.category.Name);
    const [desc, setDesc] = useState(props.category.Description);

    const deleteCategory = () => {
        if (window.confirm(`Are you sure you want to delete ${props.category.Name}?`)) {
            axios.delete('https://todoapi.jamesrcaldwell.com/api/categories/' + props.category.CategoryId).then(response => { props.getCats() })
        }
    }

    const editCat = (e) => {
        const catToEdit = {
            CategoryId: props.category.CategoryId,
            Name: name,
            Description: desc
        };

        axios.put('https://todoapi.jamesrcaldwell.com/api/categories/', catToEdit).then(() => {
            setEditMode(false);
            props.getCats();
        })
    }

    return (

        <tr>
            {!editMode ?
                <>
                    <td onClick={() => setEditMode(true)}>
                        <span>
                            <button onClick={() => setEditMode(true)} className="editCat">
                                <FontAwesomeIcon icon={['fas', 'edit']} />
                            </button>
                        </span>
                        <span data-testid="catName" className="catName">{props.category.Name}</span>
                    </td>
                    <td onClick={() => setEditMode(true)}>{props.category.Description !== null ?
                        props.category.Description :
                        'No Description Provided'}
                    </td>
                    <td>
                        <button className="deleteCat" onClick={() => deleteCategory()}>
                            <FontAwesomeIcon icon={['fas', 'trash-alt']} />
                        </button>
                    </td>
                </>
                :
                <>
                    <td>
                        <input type="text" className="form-control" defaultValue={name} onChange={(e) => setName(e.currentTarget.value)} placeholder="Name" />
                    </td>
                    <td>
                        <input type="text" className="form-control" defaultValue={desc} onChange={(e) => setDesc(e.currentTarget.value)} placeholder="Description" />
                    </td>
                    <td>
                        <button onClick={() => editCat()} className="editCat text-uppercase btn btn-warning m-1">Save</button>
                        <button onClick={() => setEditMode(false)} className="btn btn-danger m-1">Cancel</button>
                    </td>
                </>
            }
        </tr>
    )
}
