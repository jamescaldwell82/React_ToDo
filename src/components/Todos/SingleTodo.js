import React from 'react'
import axios from 'axios'
import { useAuth } from '../../contexts/AuthContext';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { library } from '@fortawesome/fontawesome-svg-core';
import { fas } from '@fortawesome/free-solid-svg-icons';

library.add(fas);

export default function SingleTodo(props) {
    const { currentUser } = useAuth();

    const toggleTodo = (done) => {
        const todo = {
            TodoId: props.todo.TodoId,
            Action: props.todo.Action,
            Done: done,
            CategoryId: props.todo.CategoryId
        }

        axios.put(`https://todoapi.jamesrcaldwell.com/api/todoitems`, todo).then(() => {
            props.getTodos();
        })

    }

    const deleteTodo = (id) => {
        if (window.confirm(`Are you sure you want to delete ${props.todo.Action}?`)) {
            axios.delete(`https://todoapi.jamesrcaldwell.com/api/todoitems/${id}`).then(() => { props.getTodos() })
        }
    }

    return (
        <div className="singleTodo">
            {currentUser.email === 'james.caldwell82@outlook.com' &&
                <button id="deleteLink" onClick={() => deleteTodo(props.todo.TodoId)}>
                    <FontAwesomeIcon icon={['fas', 'trash-alt']} />
                </button>
            }
            {currentUser.email !== 'james.caldwell82@outlook.com' ?
            <h1>{props.todo.Action}</h1> :    
            <h1>{props.todo.Action}
                {props.todo.Done === true ?
                    <span className="text-success" onClick={() => toggleTodo(!props.todo.Done)} title="Click to update status...">Complete</span> :
                    <span className="text-danger" onClick={() => toggleTodo(!props.todo.Done)} title="Click to update status...">In Progress</span>
                }
            </h1>
        }
            <p>{props.todo.Category.Name}</p>
        </div>
    )
}
