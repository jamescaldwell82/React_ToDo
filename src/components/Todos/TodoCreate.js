import React, { useRef, useState } from 'react'
import { Form, Row, Col } from 'react-bootstrap'

export default function TodoCreate(props) {
    const actionRef = useRef();
    const catRef = useRef();

    const [actionVal, setActionVal] = useState('');


    const handleSubmit = (e) => {
        e.preventDefault();

        const todo = {
            Action: actionRef.current.value,
            Done: false,
            CategoryId: catRef.current.value
        }
    
        //put code here that will validate to check on stringLength and if the same todo has been created in the same category.
        
        props.addTodo(todo);

    }
    return (
        <div className="singleTodo createTodo text-white">
            {/* <h1>Create New Task</h1>
            <hr className="bg-white" /> */}
            <Form onSubmit={handleSubmit}>
                        <Form.Group id="action">
                            <Form.Control className="text-white" type="text" ref={actionRef} placeholder="Enter task..." />
                            {actionVal !== '' &&
                                <div className="text-danger">
                                    {actionVal}
                                </div>
                            }
                            <select className="form-control" ref={catRef}>
                                <option>Select a Category...</option>
                                {props.categories.map(x => 
                                    <option key={x.CategoryId} value={x.CategoryId}>{x.Name}</option>    
                                )}
                            </select>
                        </Form.Group>

                        <Form.Group id="button">
                            <button type="submit" className="btn btn-dark">Submit</button>
                        </Form.Group>
            </Form>
        </div>
    )
}
