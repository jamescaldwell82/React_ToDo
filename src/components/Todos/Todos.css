section.todos {
    background-image: url(/src/images/background.jpg);
    background-size: 100%;
    min-height: 78vh;
}
section.todos hr {
    height: 4px;
    background-color: rgb(233,236, 22);
}
section.todos article.intro {
    border-bottom: 4px solid rgb(233,236, 22);
}
.taskGallery {
    padding: 3em 0;
}

div.singleTodo {
    width: 75%;
    padding: 1em;
    margin: 3px auto;
    border: 1px solid #5e5e5e;
    position: relative;
    background-color: rgba(255,255,255,.9);
    overflow:hidden;
}


div.singleTodo h1 {
    font-size: 2em;
    width: 80%;
}


div.singleTodo span {
    margin-left: 3em;
    font-size: .5em;
    position: absolute;
    right: 0;
    top: 0;
    bottom: 0;
    height: 100%;
    background-color: black;
    padding: 1.75em .75em;
    font-weight: bold;
    width: 25%;
    cursor: pointer;
}

article.todo-complete h2, article.todo-incomplete h2 {
    margin-bottom: 0;
}
span.todo-count {
    font-size: .5em;
    display: block;
}

div.singleTodo p {
    position: absolute;
    top: .25em;
    left:.5em;
    font-weight: bold;
    color: rgb(100,100,100); 
    font-size: .85em;
}

div.singleTodo div {
    margin: 0;
}

div.createTodo {
    padding: 0;
    position: relative;
    padding: 5px;
    background-color: black;
    border: 1px solid rgb(189, 189, 189);
}

div.createTodo div:first-child {
    width: 75%;   
}

div.createTodo div:first-child input{
    border-radius: 0;
    outline: none;
    border: none;
    background-color: transparent;
    color: rgb(233,236, 22);
    font-size:1.2em;
}

div.createTodo div:first-child hr {
    background-color: rgb(233,236, 22);
    margin:3px;
}
div.createTodo div:first-child select {
    background-color: transparent;
    border: none;
    margin-top: 0;
    color: #000;
    background-color: rgb(255,255,255);
    display:block;
    text-transform: uppercase;
    font-weight: bold;
    letter-spacing: 1px;
    position:relative;
    right: 2px;
    border-radius: 0;
}

div.createTodo div:first-child select:active, div.createTodo div:first-child select:focus {
    border: none;
}
div.createTodo div:first-child option {
    background-color: transparent;
}

div.createTodo div:first-child input::placeholder, div.createTodo div:first-child input:focus {
    color: rgb(233,236, 22);
    box-shadow: none;
}

div.createTodo div:last-child {
    width: 25%;
    position: absolute;
    right: 0;
    top: 0;
    bottom: 0;
}
div.createTodo button {
    width: 100%;
    height: 100%;
    border-radius: 0;  
    border: none;
    outline: none;
    background-color: white;
    color: black;
    font-weight: bold;
    letter-spacing: 1px;
    text-transform: uppercase;
    font-size: 1.5em;
    transition:ease-in-out .5s;
    border-left: 1px solid gray;
}

div.createTodo button:hover {
    background-color: rgb(233,236, 22);
    color: rgb(50,50,50);
    border-left: 1px solid rgb(233,236, 22)
}
.galleryButtons {
    width: 75%;
    margin: auto;
}

.galleryButtons .btn.btn-black{
    transition: .25s ease-in-out;
    background-color: white;
    color: black;
    font-weight: bold;
    letter-spacing: 1px;
}

.galleryButtons .btn.btn-black:hover {
    background-color: rgb(100,100,100);
    color: white;
}


.galleryButtons .btn {
    height: 4em;
    width: 13em;
    border-radius: 0;
}

#deleteLink {
    position: absolute;
    left: 0;
    bottom: 0;
    padding: 10px 15px;
    background-color: red;
    color: white;
    border: none;
    border-top-right-radius: 4px;
}

#deleteLink:hover {
    color: rgb(233,236, 22);
    border-color: rgb(233,236, 22);
    background-color: black;
}

@media only screen and (max-width: 1023px){
    div.singleTodo {
        width: 95%;
    }
    div.createTodo div:last-child {
        width: 25%;
    }
}


