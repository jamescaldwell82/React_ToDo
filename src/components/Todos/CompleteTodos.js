import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import React, { useState } from 'react'
import SingleTodo from './SingleTodo'

export default function CompleteTodos(props) {
    const [visibilityControl, setVisibilityControl] = useState(false);

    const getNumberOfTodos = () => {
        if (props.filter !== 0) {
            return props.todos.filter(x => x.Done === true).filter(x => x.CategoryId === props.filter).length;
        }
        else {
            return props.todos.filter(x => x.Done === true).length; 
        }
    }
    return (
        <article className="todo-complete">
            <h2 className="text-white p-4">Completed Tasks
                <button onClick={() => setVisibilityControl(!visibilityControl)} className="vControl btn btn-light m-2">
                    {visibilityControl === false ?
                        <FontAwesomeIcon icon={['fas', 'plus']} /> :
                        <FontAwesomeIcon icon={['fas', 'minus']} />}
                </button>
                <span className="todo-count">
                    ({getNumberOfTodos()} item{getNumberOfTodos() === 1 ? '' : 's'} completed)
                </span>
            </h2>

            {visibilityControl &&
                <>
                    {props.filter === 0 && props.todos.filter(x => x.Done === true).map(x =>
                        <SingleTodo
                            key={x.TodoId}
                            todo={x}
                            getTodos={props.getTodos} />
                    )}
                    {props.filter !== 0 && props.todos.filter(x => x.Done === true).filter(x => x.CategoryId === props.filter).map(x =>
                        <SingleTodo
                            key={x.TodoId}
                            todo={x}
                            getTodos={props.getTodos} />
                    )}
                </>
            }
        </article>
    )
}
