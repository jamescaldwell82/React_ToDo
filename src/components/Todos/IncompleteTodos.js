import React, { useState } from 'react'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import SingleTodo from './SingleTodo'

export default function IncompleteTodos(props) {
    const [visibilityControl, setVisibilityControl] = useState(true);

    return (
        <article className="todo-incomplete">
            <h2 className="text-white p-4">Incomplete Tasks
                <button onClick={() => setVisibilityControl(!visibilityControl)} className="vControl btn btn-light m-2">
                    {visibilityControl === false ?
                        <FontAwesomeIcon icon={['fas', 'plus']} /> :
                        <FontAwesomeIcon icon={['fas', 'minus']} />}
                </button>
                <span className="todo-count">
                    ({props.filter === 0 ?
                        props.todos.filter(x => x.Done === false).length :
                        props.todos.filter(x => x.Done === false).filter(x => x.CategoryId === props.filter).length} items to complete)
                </span>
            </h2>
            {visibilityControl &&
                <>
                    {props.filter === 0 && props.todos.filter(x => x.Done === false).map(x =>
                        <SingleTodo
                            key={x.TodoId}
                            todo={x}
                            getTodos={props.getTodos} />
                    )}
                    {props.filter !== 0 && props.todos.filter(x => x.Done === false).filter(x => x.CategoryId === props.filter).map(x =>
                        <SingleTodo
                            key={x.TodoId}
                            todo={x}
                            getTodos={props.getTodos} />
                    )}
                </>
            }
        </article>
    )
}
