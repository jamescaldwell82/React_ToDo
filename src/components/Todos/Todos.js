import React, { useState, useEffect } from 'react'
import { useAuth } from '../../contexts/AuthContext'
import { Container } from 'react-bootstrap'
import axios from 'axios'
import SingleTodo from './SingleTodo';
import './Todos.css'
import TodoCreate from './TodoCreate';
import CompleteTodos from './CompleteTodos';
import IncompleteTodos from './IncompleteTodos';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { library } from '@fortawesome/fontawesome-svg-core';
import { fas } from '@fortawesome/free-solid-svg-icons';
import CatFilter from '../Categories/CatFilter';
import CatCreate from '../Categories/CatCreate';

library.add(fas);

export default function Todos() {
    const [todos, setTodos] = useState([]);
    const [categories, setCategories] = useState([]);
    const [showCreateForm, setShowCreateForm] = useState(false);
    const { currentUser } = useAuth();
    const [filter, setFilter] = useState(0);

    const getTodos = () => {
        axios.get('https://todoapi.jamesrcaldwell.com/api/todoitems').then(response => {
            setTodos(response.data);
        })
    }

    const getCats = () => {
        axios.get('https://todoapi.jamesrcaldwell.com/api/categories').then(response => {
            setCategories(response.data);
        })
    }

    //Create
    const addTodo = (todo => {
        axios.post('https://todoapi.jamesrcaldwell.com/api/todoitems', todo).then(response => {
            getTodos();
            setShowCreateForm(false);
        })
    })

    useEffect(() => {
        getTodos();
        getCats();
    }, []);

    return (
        <section className="todos">
            <article className="intro text-white p-5">
                <h1>Tasks to Accomplish</h1>
            </article>
            <Container>
                <CatFilter categories={categories} filter={filter} setFilter={setFilter} getCats={getCats} />

                <div className="taskGallery">
                    {currentUser.email === 'james.caldwell82@outlook.com' &&
                    <>
                        <div className="galleryButtons text-right">
                            {showCreateForm === false ?
                                <button onClick={() => setShowCreateForm(true)} className="btn btn-black bold mb-3">
                                    Add New Task <FontAwesomeIcon icon={['fas', 'plus']} />
                                </button> :
                                <button onClick={() => setShowCreateForm(false)} className="btn btn-danger bold mb-3">
                                    Cancel <FontAwesomeIcon icon={['fas', 'times']} />
                                </button>
                            }
                        </div>
                            {showCreateForm &&
                                <TodoCreate
                                    categories={categories}
                                    addTodo={addTodo}
                                />
                            }
                            </>
                    }
                    <IncompleteTodos todos={todos} getTodos={getTodos} filter={filter} />
                </div>
            </Container>
            <hr />
            <Container>
                <div className="taskGallery">
                    <CompleteTodos todos={todos} getTodos={getTodos} filter={filter} />
                </div>
            </Container>

        </section>
    )
}
