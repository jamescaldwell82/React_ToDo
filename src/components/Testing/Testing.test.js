import React from 'react'
import ReactDOM from 'react-dom'
import Testing from './Testing'

//The below items are imported to bring in specific testing functionalities. First are some tools that will allow us to perform shallow rendering tests and clean up the data after the test. Second are tools to make the tests actually function
import {render, cleanup, getByTestId} from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'
import renderer from 'react-test-renderer'
import Todos from '../Todos/Todos'
import { AuthProvider } from '../../contexts/AuthContext'
import CatRow from '../Categories/CatRow'
import CompleteTodos from '../Todos/CompleteTodos'

//This will clean up the data after each test
afterEach(cleanup);

it('renders w/o crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Testing/>, div);
})


it('renders Testing correctly', () => {
    const tree = renderer.create(
            <Testing Name='Joe' 
            Desc='Test' />
    ).toJSON();

    expect(tree).toMatchSnapshot();
})

it('renders Testing Desc prop correctly', () => {
    const {getByTestId} = render(<Testing Name="James" Desc="test"/>);

    expect(getByTestId('paragraph')).toHaveTextContent('test');
})

t




