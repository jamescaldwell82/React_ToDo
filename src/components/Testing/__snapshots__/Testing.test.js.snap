// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Testing correctly 1`] = `
<section
  className="testing"
>
  <article
    className="text-center md-0 p-5"
  >
    <h1>
      Joe
    </h1>
    <p
      data-testid="paragraph"
    >
      Test
    </p>
  </article>
</section>
`;
